### ISSUE TITLE
Auto-profile loading by window title now working correctly

### ISSUE THREAD

**Issue by TheManuel2** [schmaldeo/DS4Windows#20]
**Describe the bug**
Although DS4W has supported auto-profile loading by window title for a long time, in addition to by process (.exe), it has never really worked well.  This is particularly useful for emulators where you might want to use your default profile for navigating the main interface, but then switch to a different profile when a game is loaded.  
Most emulators change their window title when a game is loaded, so it's not hard to identify a character in the title that is common among games to load a different profile, but DS4W doesn't recognize the change unless you alt+Tab out of the emulator window and then go back to it.

**To Reproduce**
1. Identify the window titles of any emulator with and without loading a game 
2. Assign an auto profile to the emulator but use the window title field and try to use one of the 3 methods (beginning of string, end of string, and contains a string) so that the profile only loads when the emulator loads a game

**Expected behavior**
Your auto-profile should load only when a game is launched, but not when the emulator is idle just showing its GUI

**Screenshots and Logs**
I'll be happy to provide any logs that might be helpful

**Desktop (please complete the following information):**
 - Controller Make and Model: Sony DS4 CUH-ZCT2U - yes, same one as in the template example :-)
 - OS: Windows 11 Pro 23H2, Build 22631.4169
 - DS4Windows Version: 3.3.3

**Additional context**
My main use case is that I often use the controller as mouse so my default profile is used by Controller Companion to navigate Windows, but when I load a game in an emulator, I switch to an auto profile that only has the escape key mapped to the guide button, because I like to use each emulator's native DS4 support.  I hope this is not too niche that it would not be worthwhile fixing the auto-profiles feature.

Thanks for picking up DS4Windows!!!

**Comment by major-sanyi** [comment]
- What emulator are you using?
- What are your Window Title parameters
- What is the exact Window titles?
     -  For example PCSX2 has the title PCSX2 vX where X is version number. And games have their own titles.
     
I will look into adding new string operator(s).

**Comment by TheManuel2** [comment]
Here are a non-working example and a working example (sort-of):

Yuzu (auto-profiles don't work):
- Window title without loading a game: "yuzu Early Access 4176"
- Window title after a game is loaded: "yuzu Early Access 4176 | Mario Kart 8 Deluxe (32-bit) | 3.0.1 | NVIDIA"
- My Window Title parameters: NVIDIA$
To save you some time, I should reiterate that if I bring a different screen into focus, say File Explorer, and then go back to the game screen the auto-profile loads, but if I always stay on the yuzu screen, DS4W never notices that the window title changed.  Perhaps the logic of parsing the window title is fine, but there is a real-time detection issue (in my poor layman's terms).

Dolphin (the "sort-of" working example):
- Window title without loading a game: "Dolphin 2409-45"
- Window title after a game is loaded: "Dolphin 2409-45 | JIT64 DC | Direct3D 11 | HLE | Mario Kart Wii (RMCE01)"
- My Window Title parameters:
- By trial and error, I made it work by creating 2 Auto Profile entries for Dolphin with these Window Title parameters:
     1st entry: "Dolphin"
     2nd entry:  *JIT64
I wish I could tell why this works.  I only included here in case it provides you a clue.

Please let me know if you need more information, and thank you again.

**Comment by schmaldeo** [comment]
I'll work on it today or tomorrow, I think I know what the issue is and how to handle it :))

**Comment by TheManuel2** [comment]
Awesome!

**Comment by schmaldeo** [comment]
It's available in [v3.8.2](https://github.com/schmaldeo/DS4Windows/releases/tag/v3.8.2). It works with what I've checked, please check if it works for you and let me know, so I can close the issue if it does :)

Technical notes:
didn't use the win event hook for it as the current implementation kinda works around what could be achieved with NT API functions. Instead I'm actually checking for title change in the function that processes current foreground window. It doesn't seem to cause any memory problems + it uses a slightly different implementation of the `GetWindowText` that uses a `char*` instead of a `StringBuilder`. The function uses stackalloc, so the memory is freed as soon as the function returns, therefore any memory leaks should not be a problem.

I also added the [`Microsoft.Windows.CsWin32`](https://github.com/microsoft/CsWin32) library and I'm planning to modify all the code that interacts with Windows APIs to use this library instead of using the `DllImport`ed functions, as in my opinion it's a massive QoL improvement for developers (provides all the types and functions that you want).

**Comment by TheManuel2** [comment]
That sounds awesome!  I’m eager to try this when I get home tonight.

**Comment by TheManuel2** [comment]
I just tested the fixes with 5 different emulators, and it works great, with one exception: Ryujinx.  

When opening the emulator's main window, the default profile stays active as it should, and when I launch a game the auto-profile kicks in correctly.  However, when stopping emulation (Esc key) and the GUI shows up again, DS4W stays stuck on the loaded auto-profile and does not return to the default one. 
I tried 4 different window title strings:
*-
*bit
)$
bit)$

With yuzu everything works beautifully when launching and sopping games.  Likewise with Dolphin, although that one always worked well.  RPCS3 and CEMU work great too, but those don't have a way to stop emulation and returning to the GUI, so I can't verify if returning from a game relinquishes the auto-profile, nor does it matter on those emulators anyway if emulation can't be stopped :-)
